src/ott/solvers/linear/univariate.py

```diff
@@ -91,12 +104,52 @@ def __call__(self, x: jnp.ndarray, y: jnp.ndarray) -> float:
       sorted_x, sorted_y = self.sort_fn(x), self.sort_fn(y)
       xx = jnp.quantile(sorted_x, q=jnp.linspace(0, 1, self.n_subsamples))
       yy = jnp.quantile(sorted_y, q=jnp.linspace(0, 1, self.n_subsamples))
+    elif self.method == "wasserstein":
+      return self._cdf_distance(x, y, a, b)
     else:
       raise NotImplementedError(f"Method `{self.method}` not implemented.")
 
     # re-scale when subsampling
     return self.cost_fn.pairwise(xx, yy) * (n / xx.shape[0])
 
+  def _cdf_distance(
+      self, x: jnp.ndarray, y: jnp.ndarray, a: Optional[jnp.ndarray],
+      b: Optional[jnp.ndarray]
+  ):
+    # Implementation based on `scipy` implementation for
+    # :func:<scipy.stats.wasserstein_distance>
+    x, a = self._validate_distribution(x, a)
+    y, b = self._validate_distribution(y, b)
+
+    all_values = jnp.concatenate([x, y])
+    all_values_sorter = jnp.argsort(all_values)
+    all_values_sorted = all_values[all_values_sorter]
+    x_pdf = jnp.concatenate([a, jnp.zeros(y.shape)])[all_values_sorter]
+    y_pdf = jnp.concatenate([jnp.zeros(x.shape), b])[all_values_sorter]
+
+    x_cdf = jnp.cumsum(x_pdf)
+    y_cdf = jnp.cumsum(y_pdf)
+
+    quantiles = jnp.sort(jnp.concatenate([x_cdf, y_cdf]))
+    x_cdf_inv = all_values_sorted[jnp.searchsorted(x_cdf, quantiles)]
+    y_cdf_inv = all_values_sorted[jnp.searchsorted(y_cdf, quantiles)]
+    return jnp.sum(
+        jax.vmap(self.cost_fn)(y_cdf_inv[1:, None], x_cdf_inv[1:, None]) *
+        jnp.diff(quantiles)
+    )
+
+  def _validate_distribution(self, x, p):
```

Comment:
I'd remove (also the shape check, would say not really necessary) this function and use this in the code above:
```python
a = jnp.ones_line(x) if a is None else a
a /= jnp.sum(a)
b = jnp.ones_like(y) if b is None else b
b /= jnp.sum(b)

```

Reply:
Sounds good to me!